VDServer overhaul

aBothe opened the pull request:
Hier schonmal ein paar erste Erweiterungen.
Es muss noch nichts gemerged werden; Ich habe erstmal einfach nur einen PR aufgemacht :)

Schau dir mal VDServerEditorFlags an - darin habe ich die ganzen Zustände verpackt.
Das ist vielleicht einfacher, zu lesen^^

rainers commented on vdc/abothe/comserver/VDServer.cs:
```diff
@@ -1,996 +1,895 @@
+		private static string normalizeDir(string dir)
+		{
+			dir = normalizePath(dir);
+			if (dir[dir.Length - 1] == Path.DirectorySeparatorChar)
```
Hier besser noch dir.Length == 0 checken?

aBothe replied:
Jo okay, wobei Path.GetFullPath das auch schon machen könnte.

rainers commented on vdc/abothe/comserver/VDServer.cs:
```diff
@@ -1,996 +1,895 @@
+			if ((_flags & 0x8000000) != 0)
+				versions += "CRuntime_Microsoft\n";
+			else if ((_flags & 0x4000040) != 0) // GNU or LDC
+				versions += "CRuntime_MinGW\n";
```
Für LDC ist das ja schon länger falsch...
Seit kurzem könnte man die flags auch ignorieren, da die Versionen alle übertragen werden: es wird der `predefs` output von `dmd -o- -v x.obj` verwendet.

aBothe replied:
Vernünftiger :sweat_smile: - aber das enum wird auch für Mixin, UFCS und HideDeprecatedNodes verwendet
-- vielleicht könnten wir ja ein paar neue Flags verbauen, wenn sich etwas sinnvolles findet :)

rainers commented:
Sieht soweit gut aus.

> Schau dir mal VDServerEditorFlags an - darin habe ich die ganzen Zustände verpackt.

Wenn die flags nur für die Versions benötigt werden, sind sie wohl leidlich obsolet. Sorry ;-)

Wegen der Appveyor-failures: Braucht es einen update des dparsers?

aBothe commented:
Ich denke, ja, ein DParser-Update wäre notwendig - hatte mich auch schon gewundert, warum der Build fehlschlägt^^

rainers commented:
> Ich denke, ja, ein DParser-Update wäre notwendig

DParser-repo sollte aktualisiert sein.

> Es muss noch nichts gemerged werden; Ich habe erstmal einfach nur einen PR aufgemacht :)

Kann inzwischen gemergt werden?

rainers commented:
BTW: kann man die Tests als Teil des Appveyor-builds laufen lassen?

aBothe commented:
> > Ich denke, ja, ein DParser-Update wäre notwendig
> 
> DParser-repo sollte aktualisiert sein.
> 
> > Es muss noch nichts gemerged werden; Ich habe erstmal einfach nur einen PR aufgemacht :)
> 
> Kann inzwischen gemergt werden?

Ja, eigentlich schon. Es ist halt zwar nur ein Zwischenstand, aber jetzt gibt es ja einige Tests.

aBothe commented:
> BTW: kann man die Tests als Teil des Appveyor-builds laufen lassen?

Das wäre klasse!^^

Commit:
Introduced EditorDataProvider for generating EditorData instances.
Introduced TooltipGenerator.

Commit:
VDServer: Reworked GetTip() & GetTipResult() to use new editorDataProvider and TooltipGenerator infrastructure.

aBothe commented:
Schau dir mal die Art und Weise an, wie ich `GetTip()` umgemodelt habe.
Ich bin mir zwar nicht sicher, ob damit das ganze Zustandsspektrum an Asynchronität abgebildet ist, aber so braucht man zumindest keine selbstgeschriebene Mutex-Logik mehr.
Plus die Tasks können unabhängig voneinander nebeneinander laufen, sodass z.B. ein Tooltip-Request keinen Expansions-Request mehr blockiert etc.

rainers commented:
> Schau dir mal die Art und Weise an, wie ich `GetTip()` umgemodelt habe.

Blick ich noch nicht vollständig durch, aber ich bin mir nicht sicher
- ob es Sinn macht, Anfragen parallel zu bearbeiten, da der User offensichtlich schon weitergetippt hat
- ob es gut ist, immer wieder ein neues EditorData zu erstellen. Bisher wurde es absichtlich erhalten, damit die darin enthaltenen Kontexte ihre gecachten Informationen weiterverwenden, solange sich der Source nicht ändert.

>> Kann inzwischen gemergt werden?
> Ja, eigentlich schon. Es ist halt zwar nur ein Zwischenstand, aber jetzt gibt es ja einige Tests.

Wäre ganz gut, die letzten beiden commits da herauszulassen, sie fangen etwas neues an, duplizieren aber noch Funktionen. Da würde ich gerne abwarten, bis auch andere Methoden diese werden.

aBothe commented:
Okay, dann eben bis vor diesen Änderungen mergen.
Diese Kontexte für LooseResolution-Operationen werden ja nicht bei allen Operationen verwendet - bzw. können gerne in den dazugehörigen Klassen recycelt werden.

rainers commented:
> Okay, dann eben bis vor diesen Änderungen mergen.

Habe ich 'mal gemergt. Mit den nuget packages ist aber auch keine echte Freude...

rainers commented:
> > BTW: kann man die Tests als Teil des Appveyor-builds laufen lassen?
> 
> Das wäre klasse!^^

Wie lasse ich denn die Tests laufen? Bei mir meldet VS bei "Run All":
```
[30/10/2018 09:25:14 Informational] ------ Run test started ------
[30/10/2018 09:25:14 Informational] Test project DParserCOMServer.Tests does not reference any .NET NuGet adapter. Test discovery or execution might not work for this project.
It's recommended to reference NuGet test adapters in each test project in the solution.
[30/10/2018 09:25:15 Informational] Test project DParserCOMServer.Tests does not reference any .NET NuGet adapter. Test discovery or execution might not work for this project.
It's recommended to reference NuGet test adapters in each test project in the solution.
[30/10/2018 09:25:15 Informational] ========== Run test finished: 0 run (0:00:01.132463) ==========
```

Commit:
Merge branch 'master' of https://github.com/rainers/visuald into feature/vdserver_overhaul

Commit:
Got first externalized Tooltip & semantic expansions generation working.
Use AbstractVDServerTask to handle common preparation and threading/cancellation infrastructure.

Commit:
Outsourced symbol definition locator function to dedicated task class.

Commit:
VDServer: Test GetReferences()

Commit:
VDServer: Moved References-Building to dedicated task/class.

Commit:
VDServer: Test GetIdentifierTypes()

Commit:
VDServer: Removed Request enum

Commit:
VDServer: Removed original completion thread infrastructure.
Put calculation of identifierTypes into separate method.

Commit:
VDServer: force-quit tasks after optional CompletionTimeout

aBothe commented:
So, jetzt habe ich alle asynchron ausgeführten Teile von VDServer in separate Tasks überführt.
Die Ermittlung der IdentifierTypes in UpdateModule ist API-technisch eine ziemliche Frickelei -- warum sollte man dem auch nicht die Chance geben, asynchron ausgeführt zu werden?

Oder kann bei GetIdentifierTypes auch ein `__pending__` returned werden?
Dann würde ich daraus auch noch einen dedizierten Task bauen.

Oh und die Bezeichnungen der von mir geschriebenen Task/Generator-Klassen sind unschön - das benenne ich noch um, mit einem adäquateren Suffix.

rainers commented on vdc/abothe/comserver/CodeSemantics/AbstractVDServerTask.cs:
```diff
@@ -0,0 +1,67 @@
+			if (_cancellation != null && !_cancellation.IsCancellationRequested)
+				_cancellation.Cancel();
+			if(_hardCancel != null && _hardCancel.IsCancellationRequested)
+				_hardCancel.CancelAfter(400);
```
Was ist der Unterschied von _cancellation und _hardCancel? Ich hab' eigentlich Timeouts abgedreht, da es m.E. nur auf die Geduld des Users ankommen sollte ("solange wird "Pending request" in der Status-Zeile angezeigt). Wenn er weiterarbeitet, stört es ja auch nicht sehr, wenn im Hintergrund irgendetwas weiterberechnet wird.

aBothe replied:
Mit `_hardCancel` soll der Task ganz abgeschossen werden, ohne 'graceful Shutdown'.
Ja, die genauen Timeouts kann man noch ändern, ich würde halt nur gern verhindern, dass irgendeine Resolution minutenlang freidreht :smile:

rainers replied:
Ah, verstehe jetzt. Das ist nur für den Cancel bei neuem Request, nicht das frühzeitige Abbrechen eines Requests selbst. Müsste es nicht `!_hardCancel.IsCancellationRequested` heissen?

rainers commented on vdc/abothe/comserver/VDServer.cs:
```diff
@@ -254,249 +235,64 @@ static string TextLocationsToIdentifierSpans(Dictionary<int, Dictionary<ISyntaxR
+					startIndex = _tipStart.Column - 1;
+					endLine = _tipEnd.Line;
+					endIndex = _tipEnd.Column - 1;
+					answer = "__pending__";
```
Bei `__pending__` können die positionen auch 0 sein, d.h. man muss sich _tipStart und _tipEnd nicht merken.

aBothe replied:
Gut, dann kann das auch weg :+1:

rainers reviewed (commented):
Wow :-)

rainers commented:
> Die Ermittlung der IdentifierTypes in UpdateModule ist API-technisch eine ziemliche Frickelei -- warum sollte man dem auch nicht die Chance geben, asynchron ausgeführt zu werden?

Ja, sollte man wohl in eine eigene Task auslagern, insbesondere da man zumindest den TypeResolve-Teil inkrementell machen will.

rainers commented:
> Oder kann bei GetIdentifierTypes auch ein `__pending__` returned werden?

Im Augenblick nicht, da es als Teil der Antworten von UpdateModule angesehen wird. Sollte aber wohl eine eigene API in IVDServer bekommen, z.B:
```
		void GetIdentiferTypes (string filename, int startLine, int startIndex, int endLine, int endIndex);
		void GetIdentifierTypesResult (out string types);
```

aBothe commented:
> Im Augenblick nicht, da es als Teil der Antworten von UpdateModule angesehen wird. Sollte aber wohl
> eine eigene API in IVDServer bekommen, z.B:

Okay, das käme dann im nächsten Schritt :)

aBothe commented:
Werden jetzt auch die Tests in Appveyor ausgeführt? :thinking:

Commit:
VDServer: While tasks are pending, tipStart/tipEnd don't have to be populated.

rainers commented:
> Werden jetzt auch die Tests in Appveyor ausgeführt?

Ja, seit ein paar Minuten:
https://ci.appveyor.com/project/rainers/visuald/builds/19935771/job/02ctjd85fqumw2t7#L1033

aBothe commented:
> > Werden jetzt auch die Tests in Appveyor ausgeführt?
> 
> Ja, seit ein paar Minuten:
> https://ci.appveyor.com/project/rainers/visuald/builds/19935771/job/02ctjd85fqumw2t7#L1033

Sehr cool, obwohl da jetzt eigentlich 6 Tests stehen müssten -- aber appveyor nimmt ja deinen master-branch, und nicht den, den ich mergen möchte^^

aBothe commented:
Eigentlich kann jetzt gemerged werden.

Commit:
VDServer: Fix CancelAfter()-call for task stopper token.

rainers commented:
> Eigentlich kann jetzt gemerged werden.

Ok, wer nichts wagt... ;-)

rainers commented:
> Ok, wer nichts wagt... ;-)

Hmmm, mit dem Refactoring scheint das Caching des Contextes scheinbar verloren gegangen. Jetzt denkt er über einen Goto-Definition immer wieder neu nach, vorher nur beim ersten mal, bis man den Source verändert. Irgendeine Idee?

Bei Manu beendet sich DParserComServer der beta3 wohl auch immer sofort, aber leider keine weiteren Infos.

aBothe commented:
Aber das hat doch in Mono-D auch immer geklappt :thinking: 
Irgendwie muss dieses Resolution-Zeug schneller werden. Weniger Visits; vielleicht sind es die vielen virtuellen Funktionen und Überladungen der Visitors, die eine Branch-Prediction der CPU unmöglich machen.

rainers commented:
> Aber das hat doch in Mono-D auch immer geklappt

Vielleicht nur auf relativ kleinen Dateien? Das Erhalten des Contextes hatte ich irgendwann in VDServer.cs eingeführt. Ich habe das jetzt mal in der AbstractVDServerTask nachgebaut: https://github.com/rainers/visuald/commit/3a091aed9ce6b6db1a77d493d52459c084168d11?w=1 Damit bleibt der Kontext zuindest pro Task erhalten, solange sich nichts am Source ändert.


> Irgendwie muss dieses Resolution-Zeug schneller werden. Weniger Visits; vielleicht sind es die vielen virtuellen Funktionen und Überladungen der Visitors, die eine Branch-Prediction der CPU unmöglich machen.

Ich kann mir eigentlich nicht vorstellen, dass es alleine an solchen "Micro-Optimierungen" liegen kann. Vielleicht noch mehr Informationen cachen?

rainers commented on vdc/abothe/comserver/VDServer.cs:
```diff
@@ -167,25 +138,36 @@ public void UpdateModule(string filename, string srcText, int flags)
-				_setupEditorData();
+				var editorData = _editorDataProvider.MakeEditorData();
+				var cancelTokenSource = new CancellationTokenSource();
+				cancelTokenSource.CancelAfter(300);
```
Ah, deswegen ist der identifier-update erträglich schnell geworden, aber leider auch recht unzuverlässig...
